diffusion/evaluation/generate_images.py

```diff
@@ -0,0 +1,177 @@
+from composer.utils import dist
+from composer.utils.file_helpers import get_file
+from composer.utils.object_store import OCIObjectStore
+from datasets import DatasetDict
+from diffusers import AutoPipelineForText2Image
+from torch.utils.data import Dataset
+from torchvision.transforms.functional import to_pil_image
+from tqdm.auto import tqdm
+
+
+class ImageGenerator:
+    """Image generator that generates images from a dataset and saves them.
+
+    Args:
+        model (torch.nn.Module): The model to evaluate.
+        dataset (Dataset): The dataset to use the prompts from.
+        load_path (str, optional): The path to load the model from. Default: ``None``.
+        local_checkpoint_path (str, optional): The local path to save the model checkpoint. Default: ``'/tmp/model.pt'``.
+        load_strict_model_weights (bool): Whether or not to strict load model weights. Default: ``True``.
+        guidance_scale (float): The guidance scale to use for evaluation. Default: ``7.0``.
+        height (int): The height of the generated images. Default: ``1024``.
+        width (int): The width of the generated images. Default: ``1024``.
+        caption_key (str): The key to use for captions in the dataloader. Default: ``'caption'``.
+        load_strict_model_weights (bool): Whether or not to strict load model weights. Default: ``True``.
+        seed (int): The seed to use for generation. Default: ``17``.
+        output_bucket (str, Optional): The remote to save images to. Default: ``None``.
+        output_prefix (str, Optional): The prefix to save images to. Default: ``None``.
+        extra_keys (list, Optional): Extra keys from the dataset to include in the metadata. Default: ``None``.
+        additional_generate_kwargs (Dict, optional): Additional keyword arguments to pass to the model.generate method.
+    """
+
+    def __init__(self,
+                 model: Union[torch.nn.Module, str],
+                 dataset: Union[Dataset, DatasetDict],
+                 load_path: Optional[str] = None,
+                 local_checkpoint_path: str = '/tmp/model.pt',
+                 load_strict_model_weights: bool = True,
+                 guidance_scale: float = 7.0,
+                 height: int = 1024,
+                 width: int = 1024,
+                 caption_key: str = 'caption',
+                 seed: int = 17,
+                 output_bucket: Optional[str] = None,
+                 output_prefix: Optional[str] = None,
+                 extra_keys: Optional[list] = None,
+                 additional_generate_kwargs: Optional[Dict] = None,
+                 hf_model: Optional[bool] = False):
+
+        if isinstance(model, str) and hf_model == False:
+            raise ValueError('Can only use strings for model with hf models!')
+        self.hf_model = hf_model
+        if hf_model or isinstance(model, str):
+            print(f'LOCALRANK{dist.get_local_rank()}')
+            if dist.get_local_rank() == 0:
```

Comment:
Is this to ensure checkpoint download only happens on local rank 0, and the subsequent `from_pretrained` call doesn't trigger another download?

Reply:
yup, also added to the dataset loading logic